src/sqlfluff/rules/L003.py

```diff
@@ -776,136 +660,358 @@ def _process_current_line(
+            template_line.line_no,
+            fixes,
+        )
+        return LintResult(
+            anchor=trigger_segment,
+            memory=memory,
+            description=_Desc(
+                len(desired_indent) // self.tab_space_size,
+                this_line.indent_size,
+                template_line.line_no,
+            ),
+            fixes=fixes,
+        )
 
-    @classmethod
-    def _single_placeholder_line(cls, current_line):
+
+class _TemplateLineInterpreter:
+    start_blocks = (
+        ("placeholder", "block_start"),
+        ("placeholder", "compound"),
+        ("placeholder", "literal"),
+        ("placeholder", "block_mid"),
+    )
+    indent_types = (
+        ("indent", None),
+        ("newline", None),
+    )
+    valid_start_combos = list(
+        itertools.product(
+            start_blocks,
+            indent_types,
+        )
+    )
+    dedent_types = (("dedent", None),)
+    end_block = (
+        ("placeholder", "block_end"),
+        ("placeholder", "compound"),
+        ("placeholder", "block_mid"),
+    )
+    valid_end_combos = list(
+        itertools.product(
+            dedent_types,
+            end_block,
+        )
+    )
+
+    def __init__(
+        self,
+        current_line: List[BaseSegment],
+        templated_file: Optional[TemplatedFile],
+    ) -> None:
+        self.current_line = [el for el in current_line if not el.is_whitespace]
+        self.templated_file = templated_file
+        self._adjacent_pairs: Optional[
+            List[Tuple[Tuple[str, Optional[str]], Tuple[str, Optional[str]]]]
+        ] = None
+
+    def is_single_placeholder_line(self):
         count_placeholder = 0
-        for seg in current_line:
+        for seg in self.current_line:
             if seg.is_code:
                 return False
             elif seg.is_type("placeholder"):
                 count_placeholder += 1
         return count_placeholder == 1
 
-    @classmethod
-    def _is_template_block_start_line(cls, current_line, templated_file):
-        def segment_info(idx: int) -> Tuple[str, Optional[str]]:
-            """Helper function."""
-            seg = current_line[idx]
-            return seg.type, cls._get_element_template_info(seg, templated_file)
-
-        if not cls._single_placeholder_line(current_line):
-            return False
-        for idx in range(1, len(current_line)):
-            if segment_info(idx - 1) in (
-                ("placeholder", "block_start"),
-                ("placeholder", "compound"),
-                ("placeholder", "block_mid"),
-            ) and segment_info(idx) == ("indent", None):
-                return True
-        return False
+    def list_segement_and_raw_segement_types(self):
+        """Yields the tuple of seg type and underlying type were applicable."""
+        for seg in self.current_line:
+            raw_seg = self.get_raw_slices(seg)
+            yield (seg.type, raw_seg[0].slice_type if raw_seg else None)
```

Comment:
Do you think this might be  clearer?
```suggestion
            if raw_seg:
                yield (seg.type, raw_seg[0].slice_type
            else:
                yield None
```

Reply:
Your suggestion isnt functionally the same. But maybe a middling var ?

Reply:
Oh yes, I misread it. Variable might be good. Python ternary expressions are sometimes great and sometimes hard to read.